src/backend/Error.h

```diff
@@ -0,0 +1,80 @@
+// Copyright 2018 The NXT Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BACKEND_ERROR_H_
+#define BACKEND_ERROR_H_
+
+#include "common/Result.h"
+
+namespace backend {
+
+    // This is the content of an error value for Error or ResultOrError, split off to its own file
+    // to avoid having all files including headers like <string> and <vector>
+    class ErrorData;
+
+    // Error and ResultOrError are meant to be used as return value for function that are not
+    // expected to, but might fail. The handling of error is potentially much slower than successes.
+    using Error = Result<void, ErrorData*>;
+
+    template <typename T>
+    using ResultOrError = Result<T, ErrorData*>;
+
+    // Returning a success is done like so:
+    //   return {}; // for Error
+    //   return SomethingOfTypeT; // for ResultOrError<T>
+    //
+    // Returning an error is done via:
+    //   return MakeError("My error message");
+    ErrorData* MakeError(const char* message);
+
+#define NXT_CONCAT1(x, y) x##y
+#define NXT_CONCAT2(x, y) NXT_CONCAT1(x, y)
+#define NXT_LOCAL_VAR NXT_CONCAT2(_localVar, __LINE__)
+
+    // When Errors aren't handled explicitly, calls to functions returning errors should be
+    // wrapped in an NXT_TRY. It will return the error if any, otherwise keep executing
+    // the current function.
+#define NXT_TRY(EXPR)                                             \
```

Comment:
A NXT_THROW macro would be good to generate the errors and append the backtrace.  As is, I don't think the backtrace goes all the way back to the intial generation.